iot-e2e-tests/common/src/main/java/com/microsoft/azure/sdk/iot/common/helpers/X509CertificateGenerator.java

```diff
@@ -0,0 +1,180 @@
+        }
+        catch (Exception e)
+        {
+            System.out.println(Tools.getStackTraceFromThrowable(e));
+        }
+    }
+
+    /**
+     * generate a new certificate
+     */
+    private void generateCertificate() throws CertificateException, NoSuchAlgorithmException, OperatorCreationException, SignatureException, NoSuchProviderException, InvalidKeyException
+    {
+        generateCertificate(null);
+    }
+
+    /**
+     * generate a new certificate using the provided common name
+     */
+    private void generateCertificate(String commonName) throws CertificateException, NoSuchAlgorithmException, OperatorCreationException, SignatureException, NoSuchProviderException, InvalidKeyException
+    {
+        KeyPairGenerator keyGen = KeyPairGenerator.getInstance("RSA");
+        keyGen.initialize(1024, new SecureRandom());
+        KeyPair keypair = keyGen.generateKeyPair();
+        this.x509Certificate = createX509CertificateFromKeyPair(keypair, commonName);
+        this.x509Thumbprint = new String(Hex.encodeHex(DigestUtils.sha(x509Certificate.getEncoded())));
+        this.publicCertificate = getPublicCertificatePem(x509Certificate);
+        this.privateKey = getPrivateKeyPem(keypair.getPrivate());
+    }
+
+    /**
+     * @return the thumbprint for the the generated x509 certificate
+     */
+    public String getX509Thumbprint()
+    {
+        return x509Thumbprint;
+    }
+
+    /**
+     * @return the complete x509 certificate that was generated
+     */
+    public X509Certificate getX509Certificate()
+    {
+        return x509Certificate;
+    }
+
+    /**
+     * @return the Pem formatted public certificate for the generated certificate key pair
+     */
+    public String getPublicCertificate()
+    {
+        return publicCertificate;
+    }
+
+    /**
+     * @return the Pem formatted private key for the generated certificate key pair
+     */
+    public String getPrivateKey()
+    {
+        return privateKey;
+    }
+
+    /**
+     * Create a new self signed x509 certificate without specifying a common name
+     */
+    private static X509Certificate createX509CertificateFromKeyPair(KeyPair keyPair) throws CertificateException, NoSuchAlgorithmException, OperatorCreationException, SignatureException, NoSuchProviderException, InvalidKeyException
+    {
+        return createX509CertificateFromKeyPair(keyPair, null);
+    }
+
+    /**
+     * Create a new self signed x509 certificate with the specified common name
+     */
+    private static X509Certificate createX509CertificateFromKeyPair(KeyPair keyPair, String commonName)
+            throws OperatorCreationException, CertificateException, InvalidKeyException, NoSuchAlgorithmException,
+            NoSuchProviderException, SignatureException
+    {
+        String issuerString = "C=US, O=Microsoft, L=Redmond, OU=Azure";
+        // subjects name - the same as we are self signed.
+        String subjectString = "C=US, O=Microsoft, L=Redmond, OU=Azure";
```

Comment:
Since they are the same, don't create 2nd string.

Reply:
They are used within different contexts though, so it doesn't make sense to just have an issuerString or just have a subjectString.

Reply:
They are exactly the same value in this method.

Reply:
But they won't always necessarily be the same value. If we ever expand this class to do multilevel certificate chains, the issuer and subject strings will be different.

Reply:
It is important to keep a clear difference between which variable is used for the subject string, and which is used for the issuer string

Reply:
In existing implementation(self signed), both certificate chains use the same value. It's going to a change requirement if multilevel certificate chains are required. And there would be another parameter will be required for the function in that case. You could rename the function to `createX509SelfSignedCertificateFromKeyPair` if you'd like to.

Reply:
Alright, I can consolidate these two variables